docs/architecture/psa-migration/strategy.md

```diff
@@ -256,6 +212,137 @@ Strategies currently (early 2022) used with each abstraction layer:
 - Cipher (G1): replace calls at each call site
 - MD (G1): replace calls at each call site
 
+
+Supporting builds with drivers without the software implementation
+==================================================================
+
+This section presents a plan towards G5: save code size by compiling out our
+software implementation when a driver is available.
+
+Additionally, we want to save code sive by compiling out the
+abstractions layers that we are not using when `MBEDTLS_USE_PSA_CRYPTO` is
+enabled (see previous section): MD and Cipher.
+
+Let's expand a bit on the definition of the goal: in such a configuration
+(driver used, software implementation and abstraction layer compiled out),
+we want:
+
+a. the library to build in a reasonably-complete configuration,
+b. with all tests passing,
+c. and no more tests skipped than the same configuration with software
+   implementation.
+
+Criterion (c) ensures not only test coverage, but that driver-based builds are
+at feature parity with software-based builds.
+
+We can roughly divide the work needed to get there in the following steps:
+
+0. Have a working driver interface for the algorithms we want to replace.
+1. Have users of these algorithms call to PSA, not the legacy API, for all
```

Comment:
Is it possible to achieve this without compromising on test coverage? Do we do this 1:1 when switching to PSA API calls?
There are two reasons I'm asking:
- I can imagine some changes in data format that would result in tests being obsolete, or requiring more testing due to data format translation;
- It would be good to specify if we require the test coverage remaining the same after each step, or maybe after the whole work on a module is done. Or maybe we can just mention that there will be a "best effort" done to ensure that no substantial tests are omitted.

Reply:
I think work on a module can't be considered complete until test coverage is up to a satisfying level - for some reasonable value of satisfying.

For example, I consider the test coverage for PK in this PR to be satisfying, as the only tests skipped are those for pkparse that depend on PKCS5 and PKCS12, which represent a small fraction of the overall tests (less than 10%), and are arguably more integration tests than unit tests (handling this kind of encrypt keys is a feature of PEM, not PK).

As a counter-example, the test coverage for RSA in this PR is not satisfying IMO with 117 / 176 tests skipped in `test_suite_rsa` vs 0 skipped in the reference config. So, I do not consider the work on RSA complete.

Another counter-example that doesn't show in this PR: actually what I first tried was to make X.509 pass with drivers. This could be achieved by disabling RSA, leaving only EC certs to test with. Turned out this caused a very large proportion of tests to be skipped (I don't remember, 60% to 90% I think) and I considered this not enough, so I realised I had to work on RSA first, at least PKCS#1 v1.5.

We could test TLS 1.2 with only the PKCS#1 v1.5 part of RSA, because that's the only version it uses, but TLS 1.3 only uses PKCS#1 v2.1, so we couldn't consider it sufficiently tested if we didn't enable PKCS#1 v2.1 in the build... and so on.

(This is one of the reason I'm pushing an update with PKCS#1 v2.1 (incomplete as unit tests are missing so far), the other reason is this is an example of how I think we can deal with dependencies inside libmbecrypto (except for the PSA core's RNG) in a backwards-compatible way.)

Reply:
I'm just afraid that we might lose this information and, since there's no step planned to analyse and make sure that the coverage is the same after all the work done (or, at least good enough) - any follow-up tasks might be lost. 
Do you think that it would be worth to add an additional analysis step after the work is done to determine if the coverage is satisfying? The document puts some pressure on `c. and no more tests skipped than the same configuration with software implementation.`, but the transition to PSA itself should also have such requirement.